### PR TITLE
zfs-mount-generator: Skip loading already loaded key

### LINKED ISSUE #9495: zfs-load-key-<pool>.service asks for passphrase when the key is already loaded
### System information

Type | Version/Name
 --- | --- 
Distribution Name	| Sabayon
Distribution Version	| weekly
Linux Kernel	| 5.3.0-sabayon (5.3.4)
Architecture	| x86_64
ZFS Version	| 0.8.2-1 (https://gitlab.com/linux-be/zfs/commits/ec6bed29)
SPL Version	| 0.8.2-1 (https://gitlab.com/linux-be/zfs/commits/ec6bed29)

### Describe the problem you're observing

1. My initramfs (`genkernel` from https://gitlab.com/linux-be/genkernel) loads the key for my root pool (`te`) and mounts the boot environment's datasets.
2. `zfs-load-key-te.service` (generated by `zfs-mount-generator`) asks for the passphrase again, even though the key is already loaded.
3. The service fails, whether the passphrase is provided or not.

### Describe how to reproduce the problem
This can be reproduced with a pool on a file as well:
```
# truncate -s100M /tmp/testpool
# touch /etc/zfs/zfs-list.cache/testpool
# zpool create -O encryption=aes-256-gcm -O keyformat=passphrase testpool /tmp/testpool
Enter passphrase: 
Re-enter passphrase: 
# systemctl daemon-reload
# systemctl start zfs-load-key-testpool
Enter passphrase for testpool: ********
Enter passphrase for testpool: ********
Enter passphrase for testpool: ********
Job for zfs-load-key-testpool.service failed because the control process exited with error code.
See "systemctl status zfs-load-key-testpool.service" and "journalctl -xe" for details.
# systemctl status zfs-load-key-testpool
● zfs-load-key-testpool.service - Load ZFS key for testpool
   Loaded: loaded (/etc/zfs/zfs-list.cache/testpool; generated)
   Active: failed (Result: exit-code) since Tue 2019-10-22 12:50:32 CEST; 13s ago
     Docs: man:zfs-mount-generator(8)
  Process: 69493 ExecStart=/bin/sh -c set -eu;count=0;while [ $$count -lt 3 ];do  systemd-ask-password --id="zfs:testpool"    "Enter passphrase for testpool:"|    /sbin/zfs load-key "testp>
 Main PID: 69493 (code=exited, status=1/FAILURE)

Oct 22 12:50:19 vozhx-te systemd[1]: Starting Load ZFS key for testpool...
Oct 22 12:50:19 vozhx-te sh[69493]: Key load error: Key already loaded for 'testpool'.
Oct 22 12:50:26 vozhx-te sh[69493]: Key load error: Key already loaded for 'testpool'.
Oct 22 12:50:29 vozhx-te sh[69493]: Key load error: Key already loaded for 'testpool'.
Oct 22 12:50:32 vozhx-te systemd[1]: zfs-load-key-testpool.service: Main process exited, code=exited, status=1/FAILURE
Oct 22 12:50:32 vozhx-te systemd[1]: zfs-load-key-testpool.service: Failed with result 'exit-code'.
Oct 22 12:50:32 vozhx-te systemd[1]: Failed to start Load ZFS key for testpool.
```

### PR DESCRIPTION
**Pull request by vozhyk-**
Don't ask for the password / try to load the key
if the key for the `encryptionroot` is already loaded.

### Motivation and Context

This resolves #9495.

### Description

Make the generated `zfs-load-key-<pool>.service` check if the `encryptionroot`'s `keystatus` is `available` (meaning that the key is loaded) and return success if it is.

### How Has This Been Tested?

I've tested the change manually on top of 0.8.2 (in a fork with `libbe` and `beadm` ported from illumos; commit https://gitlab.com/linux-be/zfs/commits/85ef14c587e7cf358a8434ec31fa9ed2799247e7) on the Sabayon system from #9495.

1. After the initramfs has loaded the key for the root pool (`te`), `zfs-load-key-te.service` does not ask for the password and does not fail:

<details>
<summary>systemctl status zfs-load-key-te.service</summary>

```
# systemctl status zfs-load-key-te.service
● zfs-load-key-te.service - Load ZFS key for te
   Loaded: loaded (/etc/zfs/zfs-list.cache/te; generated)
   Active: active (exited) since Thu 2019-11-07 14:37:45 CET; 5min ago
     Docs: man:zfs-mount-generator(8)
 Main PID: 40379 (code=exited, status=0/SUCCESS)
    Tasks: 0 (limit: 4915)
   Memory: 0B
   CGroup: /system.slice/zfs-load-key-te.service

Nov 07 14:37:45 vozhx-te systemd[1]: Starting Load ZFS key for te...
Nov 07 14:37:45 vozhx-te systemd[1]: Started Load ZFS key for te.
```

</details>

2. With a file-based pool (`testpool`), the generated `zfs-load-key-testpool.service` does not ask for the passphrase when the key is already loaded, but does ask for one and loads the key when it is not loaded:

<details>
  <summary>results with a file-based pool</summary>

```
# truncate -s100M /tmp/testpool
# touch /etc/zfs/zfs-list.cache/testpool
# zpool create -O encryption=aes-256-gcm -O keyformat=passphrase testpool /tmp/testpool

Enter passphrase: 
Re-enter passphrase: 
# systemctl daemon-reload
# systemctl start zfs-load-key-testpool
# systemctl status zfs-load-key-testpool
● zfs-load-key-testpool.service - Load ZFS key for testpool
   Loaded: loaded (/etc/zfs/zfs-list.cache/testpool; generated)
   Active: active (exited) since Thu 2019-11-07 14:40:29 CET; 4s ago
     Docs: man:zfs-mount-generator(8)
  Process: 54248 ExecStart=/bin/sh -c set -eu;keystatus="$$(/sbin/zfs get -H -o value keystatus "testpool")";[ "$$keystatus" = "unavailable" ] || exit 0;count=0;while [ $$count -lt 3 ];do >
 Main PID: 54248 (code=exited, status=0/SUCCESS)

Nov 07 14:40:29 vozhx-te systemd[1]: Starting Load ZFS key for testpool...
Nov 07 14:40:29 vozhx-te systemd[1]: Started Load ZFS key for testpool.
# zfs unmount testpool
# zfs unload-key testpool
# systemctl start zfs-load-key-testpool
# systemctl restart zfs-load-key-testpool
Enter passphrase for testpool: ********
# systemctl status zfs-load-key-testpool
● zfs-load-key-testpool.service - Load ZFS key for testpool
   Loaded: loaded (/etc/zfs/zfs-list.cache/testpool; generated)
   Active: active (exited) since Thu 2019-11-07 14:40:55 CET; 3s ago
     Docs: man:zfs-mount-generator(8)
  Process: 54433 ExecStart=/bin/sh -c set -eu;keystatus="$$(/sbin/zfs get -H -o value keystatus "testpool")";[ "$$keystatus" = "unavailable" ] || exit 0;count=0;while [ $$count -lt 3 ];do >
 Main PID: 54433 (code=exited, status=0/SUCCESS)

Nov 07 14:40:51 vozhx-te systemd[1]: Starting Load ZFS key for testpool...
Nov 07 14:40:55 vozhx-te systemd[1]: Started Load ZFS key for testpool.
```

</details>

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by vozhyk-** [comment]
I considered using `ExecCondition=` in the unit instead of adding the check to `ExecStart=`, but, according to the manpage, failure of the check would mean the unit is not relevant for the running system, and the unit might get "garbage-collected". This is undesirable, as it might become useful once e.g. the user unloads the key for the dataset.

**Comment by behlendorf** [comment]
@aerusso you may want to have a look at this as well.

**Comment by aerusso** [comment]
This looks good to me. I waffled back and forth about testing for `unavailable` vs `available`, i.e., maybe test for 

    "[ \"\$\$keystatus\" = \"unavailable\" ] || exit 0;"\

Which more narrowly proceeds only if it can actually do something about it, and **succeeds** otherwise (also, a cursory glance through `man zfs` doesn't indicate when `none` will be returned---is  that referring to the `-` returned for unencrypted datasets?).

**Comment by behlendorf** [comment]
@vozhyk- when you get a chance can you update the dracut script with a similar fix.

**Comment by vozhyk-** [comment]
> ```
> "[ \"\$\$keystatus\" = \"unavailable\" ] || exit 0;"\
> ```

@aerusso Indeed, this makes more sense. This should prevent trying to load the key for an unencrypted dataset (which would also fail) if the unit is ever generated for one.

> @vozhyk- when you get a chance can you update the dracut script with a similar fix.

@rlaager @behlendorf I've added similar checks to all the places `dracut` and `initramfs` tries to load a key. Always doing the check is useful, as the user might have loaded the key manually or by other means before the script gets called.

That said, I cannot test the `dracut` and `initramfs` changes, as all of my systems use `genkernel` for the initramfs.

File: contrib/dracut/90zfs/zfs-load-key.sh.in [hunk 1]
```diff
@@ -38,6 +38,9 @@ if [ "$(zpool list -H -o feature@encryption $(echo "${BOOTFS}" | awk -F\/ '{prin
     # if the root dataset has encryption enabled
     ENCRYPTIONROOT=$(zfs get -H -o value encryptionroot "${BOOTFS}")
     if ! [ "${ENCRYPTIONROOT}" = "-" ]; then
+	KEYSTATUS="$(zfs get -H -o value keystatus "${ENCRYPTIONROOT}")"
```
Comment (rlaager) [inline contrib/dracut/90zfs/zfs-load-key.sh.in]: This file uses spaces. Can you change these from tabs to 8 spaces (in this file only)?

**Comment by vozhyk-** [comment]
Can this be added to `0.8-release`?

**Pull request merged**